jre_emul/Classes/com/google/j2objc/security/cert/IosX509Certificate.java

```diff
@@ -52,19 +62,49 @@
 public class IosX509Certificate extends X509Certificate {
 
   private long secCertificateRef;
+  
+  private Certificate certificate;
+  private TBSCertificate tbsCert;
+  private Extensions extensions;
 
   public IosX509Certificate(long secCertificateRef) {
     this.secCertificateRef = secCertificateRef;
   }
+  
+  /**
+   * This implementation is modelled after harmony's X509CertImpl
+   */
+  public void lazyDecoding() {
+    if (this.certificate == null)
+    {
+      try {
+        // decode the Certificate object
+        this.certificate = (Certificate) Certificate.ASN1.decode(new ByteArrayInputStream(getEncoded()));
+        // cache the values of TBSCertificate and Extensions
+        this.tbsCert = certificate.getTbsCertificate();
+        this.extensions = tbsCert.getExtensions();
+      } catch (CertificateEncodingException e) {
+        throw new RuntimeException(e);
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+    }
+  }
 
-/*-[
+// Moved to native finalize()
```

Comment:
My mistake, what you wrote is fine. I was confused by the commented-out dealloc -- just delete that comment and leave your native finalize method as is. Normally you can avoid finalize methods by declaring the native field as a Java Object, but here there's